src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,202 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size = 0;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (size == 0) {
+            head = newNode;
+        } else { //addAsTail
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("you use invalid index "
+                    + index + ", size = " + size);
+        } else if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            head.prev = newNode;
+            newNode.next = head;
+            head = newNode;
+            size++;
+        } else {
+            Node nodeByIndex = getNode(index);
+            Node prevNode = nodeByIndex.prev;
+            Node nextNode = nodeByIndex;
+            newNode.prev = prevNode;
+            newNode.next = nextNode;
+            prevNode.next = newNode;
+            nextNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("you use invalid index "
+                    + index + ", size = " + size);
+        }
+        Node nodeByIndex = getNode(index);
+        return (T) nodeByIndex.value;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> newNode = new Node<>(null, value, null);
+        Node previousNode;
+        if (index == 0) { //setAsHead
+            previousNode = head;
+
+            newNode.next = head.next;
+            head.next.prev = newNode;
+
+            head = newNode;
+        } else if (index == size - 1) { //addAsTail
+            previousNode = tail;
+
+            newNode.prev = tail.prev;
+            tail.prev.next = newNode;
+
+            tail = newNode;
+        } else {
+            previousNode = getNode(index);
+            Node prevNode = previousNode.prev;
+            Node nextNode = previousNode.next;
+            newNode.prev = prevNode;
+            newNode.next = nextNode;
+            prevNode.next = newNode;
+            nextNode.prev = newNode;
+        }
+        return (T) previousNode.value;
```

Comment:
Use get method to take a value without casing

Reply:
than the code does not compile